Document/0x04e-Testing-Authentication-and-Session-Management.md

```diff
@@ -296,27 +298,55 @@ Refer to the framework documentation for examples of session timeout configurati
-2. Perform a couple of operations that require authentication.
-3. Leave the session idle until it expires. After session expiry, attempt to use the same session ID to access authenticated functionality.
+The purpose of this test case is verifying logout functionality and determining whether it effectively terminates the session on both client and server and invalidates a stateless token.
+
+Failing to destroy the server-side session is one of the most common logout functionality implementation errors. This error keeps the session or token alive, even after the user logs out of the application. An attacker who gets valid authentication information can continue to use it and hijack a user's account.
+
+Many mobile apps don't automatically log users out because it is inconvenient for customers by implementing stateless authentication. The application should still have a logout function, and it should be implemented according to best practices, destroying the access and refresh token on the client and server. Otherwise, authentication can be bypassed when the refresh token is not invalidated.
+
+#### Static Analysis
+
+If server code is available, make sure logout functionality terminates the session correctly. This verification will depend on the technology. Here are different examples of session termination for proper server-side logout:
+
+- [Spring (Java)](https://docs.spring.io/autorepo/docs/spring-security/4.1.x/apidocs/org/springframework/security/web/authentication/logout/SecurityContextLogoutHandler.html "Spring (Java)")
+- [Ruby on Rails](https://guides.rubyonrails.org/security.html "Ruby on Rails")
+- [PHP](https://php.net/manual/en/function.session-destroy.php "PHP")
+
+If access and refresh tokens are used with stateless authentication, they should be deleted from the mobile device. The [refresh token should be invalidated on the server](https://auth0.com/blog/blacklist-json-web-token-api-keys/ "Blacklisting JSON Web Token API Keys").
+
+#### Dynamic Analysis
+
+Use an interception proxy for dynamic application analysis and execute the following steps to check whether the logout is implemented properly:
+
+1. Log in to the application.
+2. Access a resource that requires authentication, typically a request for private information belonging to your account.
+3. Log out of the application.
+4. Try to access the data again by resending the request from step 2.
+
+If the logout is correctly implemented on the server, an error message or redirect to the login page will be sent back to the client. On the other hand, if you receive the same response you got in step 2, the token or session ID is still valid and hasn't been correctly terminated on the server.
+The OWASP Web Testing Guide ([OTG-SESS-006](https://www.owasp.org/index.php/Testing_for_logout_functionality_%28OTG-SESS-006%29 "OTG-SESS-006")) includes a detailed explanation and more test cases.
 
-##### Verifying that 2FA is Enforced
+### Verifying that 2FA is Enforced (MSTG‑AUTH‑9)
```

Comment:
I think I see two options here: 

a. move this above to "Two-Factor Authentication and Step-up Authentication" and just make a brief intro up there.
b. move that from above here and maybe include also the "Step-up Authentication" on the same section

Also maybe consider calling the section "Testing Two-Factor Authentication" ? (+ Step-up if needed). This way all sections start with "Testing..."

Reply:
Good idea. I merged it now with Step-up authentication and rewrote the Two-Factor Authentication test case. 
Please have a look (line 311 onwards)